### DISCUSSION TITLE
How best to solve a cyclic dependency during package installation

### DISCUSSION THREAD

**Discussion by dfremont** [tox-dev/tox#3273] [Q&A] [answered]
I'm using Tox to test a package let's call `A` (for clarity), which has an optional dependency on package `B`, which in turn depends on `A`. One of my Tox environments installs `A` using an extra that also pulls in `B`. While `pip install A[extra]` works just fine, as `pip` knows how to deal with the cyclic dependency, Tox's 2-step method of first installing `A`'s dependencies and then using a second `pip` command for `A` itself doesn't always work. In particular, the following can happen:

1. During the `install_package_deps` stage, Tox uses `pip` to install `B`, and `pip` starts searching PyPI for a compatible version of `A` to install (as it doesn't know that Tox will separately install `A` later).
2. If the dependencies of the version of `A` I'm developing have changed, there may be no compatible versions on PyPI already, so the installation fails.

I'm currently working around this issue by using a Tox plugin to add an additional installation stage which installs `A` (with the `--no-deps` option) *before* the `install_package_deps` stage, so that during that stage `pip`'s resolver will know the newer version of `A` is available:
<details>
<summary>Plugin code</summary>

```python
import tox.plugin

@tox.plugin.impl
def tox_on_install(tox_env, arguments, section, of_type):
    if section == "RunToxEnv" and of_type == "package":
        assert len(arguments) == 1
        package = arguments[0]
        path = package.path
        install_args = ["--force-reinstall", "--no-deps", str(path)]
        tox_env.installer._execute_installer(install_args, "package_early")
```

</details>

This approach is fragile (I'm using some undocumented Tox APIs), and still has the issue that `A` and `B` are installed in separate `pip` commands, so dependency conflicts could be missed.

My questions are:
1. Why doesn't Tox use a single `pip` command to install both the built package and its dependencies? That would solve the cyclic dependency problem.
2. Assuming there's a good reason for the current behavior, is there a more robust way for me to work around it? I went the plugin route after not finding any configuration options that did what I needed, but I might have missed something.

Thanks in advance!

**Comment by gaborbernat** [comment, marked as answer]
The plugin route is supported, just under documented 👍

**Reply by dfremont** [reply]
Thanks, Bernát. Is there any chance of an option for Tox to install the built package and its dependencies in a single `pip` call? I imagine there's some reason why those are done separately right now.

**Reply by gaborbernat** [reply]
Likely will not happen. Just not use deps?
